system/src/system_power_manager.cpp

```diff
@@ -675,44 +699,55 @@ int PowerManager::setConfig(const hal_power_config* conf) {
   return ret;
 }
 
-void PowerManager::loadConfig() {
+int PowerManager::getConfig(hal_power_config* conf) {
   // Ignore errors since we are anyway sanitizing the config and setting some sane defaults
-  config_.size = sizeof(config_);
-  int err = hal_power_load_config(&config_, nullptr);
+  conf->size = sizeof(*conf);
```

Comment:
I can add an extra argument to specify target size `int PowerManager::getConfig(hal_power_config* conf, size_t size)`

Reply:
There is no point in an extra argument. The structure itself already has a size field. It's a convention that the caller needs to specify the size it has allocated (which in our situation may be less than the current size of this structure, again given the fact that the app may be built for 1.0.0 and the system part currently on the device is already 2.0.0 where the structure has been extended).

The callee does not modify the size and uses it as a reference to know how much it can read or write out of the provided structure.

Reply:
How did `config_ = defaultPowerConfig;` work in the original code before my changes?

Reply:
@eberseth There is a difference if we are within 1 module, or something flows across the dynalib boundary. We don't need to worry about the sizes or anything if we are within the same module. The size of this structure is universal, because its definition was the same when building such module.

When calling something through dynalib it's not guaranteed. The application allocates the structure based on the definition it has in its version of Device OS, which may not have some of the changes present in newer version, but the system part may already have them.

```c++
// 1.0.0
struct SomeStruct {
    uint16_t size;
    uint32_t field1;
    uint32_t field2;
    uint32_t field3;
    uint32_t field4;
}; // 18 bytes

// 2.0.0
struct SomeStruct {
    uint16_t size;
    uint32_t field1;
    uint32_t field2;
    uint32_t field3;
    uint32_t field4;

    uint32_t field5;
    uint32_t field6;
    uint32_t field7;
    uint32_t field8;
}; // 34 bytes

// 1.0.0 application will allocate only 18 bytes based on its definition available in its sources and will indicate that the struct passed through dynalib is indeed only 18 bytes
// 2.0.0 system code will need to be careful with the size when both reading and writing to such structs that come through dynalib, as its own definition of this same struct has already grown to 34 bytes
```